ipa-core/src/protocol/boolean/saturating_sum.rs

```diff
@@ -1,340 +0,0 @@
-use crate::{
-    error::Error,
-    ff::{Field, Gf2},
-    protocol::{boolean::or::or, context::Context, step::BitOpStep, BasicProtocols, RecordId},
-    secret_sharing::{BitDecomposed, Linear as LinearSecretSharing, LinearRefOps},
-};
-
-#[derive(Debug)]
-pub struct SaturatingSum<S: LinearSecretSharing<Gf2>> {
-    pub sum: BitDecomposed<S>,
-    pub is_saturated: S,
-}
-
-impl<S: LinearSecretSharing<Gf2>> SaturatingSum<S> {
-    pub fn new(value: BitDecomposed<S>, is_saturated: S) -> SaturatingSum<S> {
-        SaturatingSum {
-            sum: value,
-            is_saturated,
-        }
-    }
-
-    ///
-    /// # Errors
-    /// If one of the multiplications errors
-    ///
-    /// # Panics
-    /// If something try to add a bit decomposed value larger than this `SaturatingSum` can accommodate
-    pub async fn add<C>(
-        &self,
-        ctx: C,
-        record_id: RecordId,
-        value: &BitDecomposed<S>,
-    ) -> Result<SaturatingSum<S>, Error>
-    where
-        C: Context,
-        S: LinearSecretSharing<Gf2> + BasicProtocols<C, Gf2>,
-        for<'a> &'a S: LinearRefOps<'a, S, Gf2>,
-    {
-        assert!(self.sum.len() >= value.len());
-
-        let mut output_sum = Vec::with_capacity(self.sum.len());
-        let mut carry_in = S::ZERO;
-        let zero = S::ZERO;
-        for i in 0..self.sum.len() {
-            let c = ctx.narrow(&BitOpStep::from(i));
-            // When adding a value with fewer bits than the saturating sum can express
-            // we still must compute the carries, which still requires a single multiplication
-            // so there is no savings
-            let x = value.get(i).unwrap_or(&zero);
-            let sum_bit = one_bit_adder(c, record_id, x, &self.sum[i], &mut carry_in).await?;
-
-            output_sum.push(sum_bit);
-        }
-        let is_saturated = or(
-            ctx.narrow(&BitOpStep::from(self.sum.len())),
-            record_id,
-            &carry_in,
-            &self.is_saturated,
-        )
-        .await?;
```

Comment:
Whoops. This was important functionality which was lost. This is the cause of issue #863